### PR TITLE
Parser substitution (Dollar Bracket Expressions)

### LINKED ISSUE #5649: PBS option to append string to output file names
We'd like to be able to use "./run_tests --pbs" where it appends a string to all output file names.  Something like "./run_tests --pbs fred --pbs_out_tag my_string" would result in the typical Exodus output file being named <baseNameFromInputFile>_my_string.e.

### PR DESCRIPTION
**Pull request by permcody**
Enable DBE expansion support in MOOSE

closes #5649

### CODE REVIEW

File: test/tests/parser/param_substitution/param_substitution.i [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  petsc_options_value = 'hypre boomeramg'
+[]
+
+[Outputs]
+  exodus = true
+
+  # Here we use the GetPot "DBE" function to perform a substitution.
+  # The parameter "FOOBAR" can either exist in this file or
+  # be provided on the CLI
+  file_base = out_${FOOBAR}
```
Comment (jwpeterson) [inline test/tests/parser/param_substitution/param_substitution.i]: Do you really want it to be called FOOBAR?

Comment (permcody) [inline test/tests/parser/param_substitution/param_substitution.i]: Probably not...

**Comment by jwpeterson** [comment]
FYI: we've decided to merge #5794 and rebase this one before merging...

**Comment by permcody** [comment]
Yes, that's fine.

**Pull request merged**